kq_epic_post.md

```diff
@@ -0,0 +1,160 @@
+
+#### Obstacle 1: When is a state not a state
+An issue can be one of two states - open or closed. The same is true for pull requests at the database level. However at the model level, a pull request can have three states: open, closed, or merged, as determined by the [PullRequest#state method](https://github.com/github/github/blob/a18286f131f4dc80dcbab605938125fcd9ac509d/packages/pull_requests/app/models/pull_request.rb#L650-L654). The existence of this method blocked us from simply adding a `pull_requests.state` field. We briefly explored some workarounds for this naming conflict, however, the best  solution ended up being the simplest one; choose a different column name. With `pull_requests.status` in place, we could continue on our way.
+
+#### Obstacle 2: We've had one, yes, but what about second join
+While the majority of the queries had a single join to issues for the state, there were some that were a bit more greedy. These came from GraphQL calls looking for a list of pull requests by label, and they required multiple joins to form a query that looked something like this:
+
+```
+SELECT
+    `pull_requests`.`id`
+  FROM
+    `pull_requests`
+    , `issues`
+    , `issues_labels`
+    , `labels`
+  WHERE
+    `labels`.`lowercase_name` IN (
+      . . .
+    )
+    AND `pull_requests`.`repository_id` = ?
+    AND (
+      `pull_requests`.`user_hidden` = ?
+      OR `pull_requests`.`user_id` = ?
+    )
+    AND `issues`.`state` = ?
+    AND `issues`.`pull_request_id` = `pull_requests`.`id`
+    AND `issues`.`repository_id` = `pull_requests`.`repository_id`
+    AND `issues_labels`.`issue_id` = `issues`.`id`
+    AND `labels`.`id` = `issues_labels`.`label_id`
+    AND `labels`.`repository_id` = `issues`.`repository_id`
+  ORDER BY
+    `pull_requests`.`id` DESC LIMIT ?
+```
+
+Simply swapping `issues.state` for `pull_request.status` wasn't enough for these queries; they were still timing out. Thus we explored other optimizations, including breaking out subqueries, forcing or ignoring an index, and reorganizing the query structure, none of which showed improvements[^1]. We even considered storing label ids in a new column on `issues` to eliminate the extra joins, but that presented its own big lift that seemed far outside the scope of this epic.
+
+So, like the dwarves in the mines of Moria, we dug deeper to determine _who_ was hitting these queries. Over a period of 30 days, we found that between the two queries, all calls were coming from only 4 users within 4 repos. More interestingly, all 4 users were GitHub apps mimicking merge queue functionality. One in particular was causing around 10k events routinely every 20 minutes. With this new information, our minds turned towards rate limiting. Unfortunately, rate limits are set based on individual app installations and cannot be applied by query fingerprint, so they were too blunt to deploy here.
```

Comment:
> So, like the dwarves in the mines of Moria,

I am 💯 here for these literary references. I don't get them because I'm not nearly as well-read as you but +1

Reply:
😆 I considered adding a disclaimer at the top like "Warning: Lots of Lord of the Rings references incoming". 💍